h2o-algos/src/main/java/hex/ContributionsWithBackgroundFrameTask.java

```diff
@@ -0,0 +1,90 @@
+  final Frame _backgroundFrame;
+  final boolean _aggregate;
+
+  final boolean _isFrameBigger;
+
+  public ContributionsWithBackgroundFrameTask(Frame fr, Frame backgroundFrame, boolean perReference) {
+    _frame = fr;
+    _backgroundFrame = backgroundFrame;
+    assert _frame.numRows() > 0 : "Frame has to contain at least one row.";
+    assert _backgroundFrame.numRows() > 0 : "Background frame has to contain at least one row.";
+    _isFrameBigger = fr.numRows() > backgroundFrame.numRows();
+    _aggregate = !perReference;
+  }
+
+
+  @Override
+  public void map(Chunk[] cs, NewChunk[] ncs) {
+    Frame smallerFrame = _isFrameBigger ? _backgroundFrame : _frame;
+    for (int sfIdx = 0; sfIdx < smallerFrame.numRows(); ) {
+      int finalSfIdx = sfIdx;
+      Chunk[] sfCs = IntStream
+              .range(0, smallerFrame.numCols())
+              .mapToObj(col -> smallerFrame.vec(col).chunkForRow(finalSfIdx))
+              .toArray(Chunk[]::new);
+      NewChunk[] ncsSlice = Arrays.copyOf(ncs, ncs.length - 2);
+      if (_isFrameBigger) {
+        map(cs, sfCs, ncsSlice);
+        for (int i = 0; i < cs[0]._len; i++) {
+          for (int j = 0; j < sfCs[0]._len; j++) {
+            ncs[ncs.length - 2].addNum(cs[0].start() + i); // row idx
+            ncs[ncs.length - 1].addNum(sfCs[0].start() + j); // background idx
+          }
+        }
+      } else {
+        map(sfCs, cs, ncsSlice);
+        for (int i = 0; i < sfCs[0]._len; i++) {
+          for (int j = 0; j < cs[0]._len; j++) {
+            ncs[ncs.length - 2].addNum(sfCs[0].start() + i); // row idx
+            ncs[ncs.length - 1].addNum(cs[0].start() + j); // background idx
+          }
+        }
+      }
+      sfIdx += sfCs[0]._len;
+    }
+  }
+
+  abstract protected void map(Chunk[] cs, Chunk[] bgCs, NewChunk[] ncs);
+
+  // takes care of mapping over the bigger frame 
+  public Frame runAndGetOutput(Job j, Key<Frame> destinationKey, String[] names) {
+    String[] namesWithRowIdx = new String[names.length + 2];
+    System.arraycopy(names, 0, namesWithRowIdx, 0, names.length);
+    namesWithRowIdx[names.length] = "RowIdx";
+    namesWithRowIdx[names.length + 1] = "BackgroundRowIdx";
+    Key<Frame> individualContributionsKey = _aggregate ? Key.make(destinationKey + "_individual_contribs") : destinationKey;
+    Frame indivContribs = withPostMapAction(JobUpdatePostMap.forJob(j))
+            .doAll(namesWithRowIdx.length, Vec.T_NUM, _isFrameBigger ? _frame : _backgroundFrame)
+            .outputFrame(individualContributionsKey, namesWithRowIdx, null);
+    if (!_aggregate)
+      return indivContribs;
+    try {
+      return new ContributionsMeanAggregator((int) _frame.numRows(), names.length, (int)_backgroundFrame.numRows())
+              .withPostMapAction(JobUpdatePostMap.forJob(j))
+              .doAll(names.length, Vec.T_NUM, indivContribs)
+              .outputFrame(destinationKey, names, null);
+    } finally {
+      indivContribs.delete(true);
+    }
+  }
```

Comment:
I'm thinking how to make this better. So far I came up with decreasing the need of the intermediate memory usage by calculating all baselines for a set of rows (chunk) and reducing it so I don't have a frame with `frame.nrows*bgFrame.nrows` but at most `frame.nrows + chunk.nrows*bgFrame.nrows`. I tried to more than one chunk but it's hard to find a threshold when H2O starts swapping and using just one chunk scaled roughly linearly on my laptop... (I will have to do more experiments)

Do you have some better ideas how to make it more scalable @wendycwong (or anybody else)?
The computation works by calculating contribution for each row from frame against each row from bgFrame which leads to big frames and if we want Marginal SHAP, then we are just looking for an average across the whole bgFrame. 

Changes I have for the aforementioned behavior (I'm still not quite sure about the code so I keep it as a diff in this comment):
```diff
package hex;
 
 import water.*;
-import water.fvec.Chunk;
-import water.fvec.Frame;
-import water.fvec.NewChunk;
-import water.fvec.Vec;
+import water.fvec.*;
+import water.util.Log;
 
 import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
 import java.util.stream.IntStream;
 
 /***
@@ -18,7 +18,10 @@ public abstract class ContributionsWithBackgroundFrameTask<T extends Contributio
   final Frame _backgroundFrame;
   final boolean _aggregate;
 
-  final boolean _isFrameBigger;
+  boolean _isFrameBigger;
+
+  long _startRow;
+  long _endRow;
 
   public ContributionsWithBackgroundFrameTask(Frame fr, Frame backgroundFrame, boolean perReference) {
     _frame = fr;
@@ -27,14 +30,24 @@ public abstract class ContributionsWithBackgroundFrameTask<T extends Contributio
     assert _backgroundFrame.numRows() > 0 : "Background frame has to contain at least one row.";
     _isFrameBigger = fr.numRows() > backgroundFrame.numRows();
     _aggregate = !perReference;
+    _startRow = -1;
+    _endRow = -1;
   }
 
 
   @Override
   public void map(Chunk[] cs, NewChunk[] ncs) {
     Frame smallerFrame = _isFrameBigger ? _backgroundFrame : _frame;
-    for (int sfIdx = 0; sfIdx < smallerFrame.numRows(); ) {
-      int finalSfIdx = sfIdx;
+
+    long sfIdx = 0;
+    long maxSfIdx = smallerFrame.numRows();
+    if (!_isFrameBigger && _startRow != -1 && _endRow != -1) {
+      sfIdx = _startRow;
+      maxSfIdx = _endRow;
+    }
+
+    while (sfIdx < maxSfIdx) {
+      long finalSfIdx = sfIdx;
       Chunk[] sfCs = IntStream
               .range(0, smallerFrame.numCols())
               .mapToObj(col -> smallerFrame.vec(col).chunkForRow(finalSfIdx))
@@ -99,7 +112,8 @@ public abstract class ContributionsWithBackgroundFrameTask<T extends Contributio
   }
 
   abstract protected void map(Chunk[] cs, Chunk[] bgCs, NewChunk[] ncs);
-
+  
+  /*
   // takes care of mapping over the bigger frame 
   public Frame runAndGetOutput(Job j, Key<Frame> destinationKey, String[] names) {
     double reqMem = estimateRequiredMemory(names.length + 2);
@@ -129,5 +143,108 @@ public abstract class ContributionsWithBackgroundFrameTask<T extends Contributio
     } finally {
       indivContribs.delete(true);
     }
+  } */
+
+
+  void setChunkRange(int startCIdx, int endCIdx) {
+    Log.info("SHAP Setting Chunk Range from " + startCIdx + " to " + endCIdx + ".");
+    _startRow = _frame.anyVec().chunkForChunkIdx(startCIdx).start();
+    _endRow = _frame.anyVec().chunkForChunkIdx(endCIdx).start() + _frame.anyVec().chunkForChunkIdx(endCIdx)._len;
+  }
+
+
+  // takes care of mapping over the bigger frame 
+  public Frame runAndGetOutput(Job j, Key<Frame> destinationKey, String[] names) {
+    double reqMem = estimateRequiredMemory(names.length + 2);
+    double reqPerNodeMem = estimatePerNodeMinimalMemory(names.length + 2);
+
+    String[] namesWithRowIdx = new String[names.length + 2];
+    System.arraycopy(names, 0, namesWithRowIdx, 0, names.length);
+    namesWithRowIdx[names.length] = "RowIdx";
+    namesWithRowIdx[names.length + 1] = "BackgroundRowIdx";
+    Key<Frame> individualContributionsKey = _aggregate ? Key.make(destinationKey + "_individual_contribs") : destinationKey;
+
+    if (!_aggregate) {
+      if (!enoughMinMemory(reqPerNodeMem)) {
+        throw new RuntimeException("Not enough memory. Estimated minimal total memory is " + reqMem + "B. " +
+                "Estimated minimal per node memory (assuming perfectly balanced datasets) is " + reqPerNodeMem + "B. " +
+                "Node with minimum memory has " + minMemoryPerNode() + "B. Total available memory is " + totalFreeMemory() + "B."
+        );
+      }
+      Frame indivContribs = withPostMapAction(JobUpdatePostMap.forJob(j))
+              .doAll(namesWithRowIdx.length, Vec.T_NUM, _isFrameBigger ? _frame : _backgroundFrame)
+              .outputFrame(individualContributionsKey, namesWithRowIdx, null);
+
+      return indivContribs;
+    } else {
+      if (!enoughMinMemory(reqPerNodeMem)) {
+        if (minMemoryPerNode() < 5 * (names.length + 2) * _frame.numRows() * 8) {
+          throw new RuntimeException("Not enough memory. Estimated minimal total memory is " + reqMem + "B. " +
+                  "Estimated minimal per node memory (assuming perfectly balanced datasets) is " + reqPerNodeMem + "B. " +
+                  "Node with minimum memory has " + minMemoryPerNode() + "B. Total available memory is " + totalFreeMemory() + "B."
+          );
+        }
+        // Split the _frame in subsections and calculate baselines (expand the frame) and then the average (reduce the frame sieze)
+        int nChunks = _frame.anyVec().nChunks();
+        // last iteration we need memory for ~whole aggregated frame + expanded subframe
+        int nSubFrames = (int) Math.ceil(2*reqMem / (minMemoryPerNode() - 8 * _frame.numRows() * (names.length)));
+        nSubFrames = nChunks;
+        int chunksPerIter = (int) Math.max(1, Math.floor(nChunks / nSubFrames));
+
+        Log.warn("Not enough memory to calculate SHAP at once. Calculating in " + (nSubFrames) + " iterations.");
+        _isFrameBigger = false; // ensure we map over the BG frame so we can average over the results properly;
+        Frame result = new Frame(destinationKey);
+        List<Frame> subFrames = new LinkedList<Frame>();
+        try {
+          for (int i = 0; i < nSubFrames; i++) {
+            setChunkRange(i * chunksPerIter, Math.min(nChunks - 1, (i + 1) * chunksPerIter - 1));
+            Log.info("Calculating SHAP for rows from " + _startRow + " to " + _endRow + ".");
+            Frame indivContribs = clone().withPostMapAction(JobUpdatePostMap.forJob(j))
+                    .doAll(namesWithRowIdx.length, Vec.T_NUM, _backgroundFrame)
+                    .outputFrame(Key.make(destinationKey + "_individual_contribs_" + i), namesWithRowIdx, null);
+
+            subFrames.add(new ContributionsMeanAggregator((int) (_endRow - _startRow), names.length, (int) _backgroundFrame.numRows())
+                    .setStartIndex((int) _startRow)
+                    .withPostMapAction(JobUpdatePostMap.forJob(j))
+                    .doAll(names.length, Vec.T_NUM, indivContribs)
+                    .outputFrame(Key.make(destinationKey + "_part_" + i), names, null));
+            indivContribs.delete();
+          }
+          
+          Futures fs = new Futures();
+          AppendableVec[] avs = new AppendableVec[subFrames.get(0).numCols()];
+          for (int i = 0; i < avs.length; i++) {
+            avs[i] = new AppendableVec(Vec.newKey(), Vec.T_NUM);
+            NewChunk nc = avs[i].chunkForChunkIdx(0);
+            for(Frame fr : subFrames) {
+              for (int k = 0; k < fr.numRows(); k++) {
+                nc.addNum(fr.vec(i).at(k));
+              }
+            }
+            nc.close(fs);
+            result.add(subFrames.get(0)._names[i], avs[i].close(avs[i].compute_rowLayout(), fs));
+          }
+          fs.blockForPending();
+          DKV.put(result);
+          return result;
+        } finally {
+          for (Frame fr : subFrames) {
+            Frame.deleteTempFrameAndItsNonSharedVecs(fr, result);
+          }
+        }
+      } else {
+        Frame indivContribs = withPostMapAction(JobUpdatePostMap.forJob(j))
+                .doAll(namesWithRowIdx.length, Vec.T_NUM, _isFrameBigger ? _frame : _backgroundFrame)
+                .outputFrame(individualContributionsKey, namesWithRowIdx, null);
+        try {
+          return new ContributionsMeanAggregator((int) _frame.numRows(), names.length, (int) _backgroundFrame.numRows())
+                  .withPostMapAction(JobUpdatePostMap.forJob(j))
+                  .doAll(names.length, Vec.T_NUM, indivContribs)
+                  .outputFrame(destinationKey, names, null);
+        } finally {
+          indivContribs.delete(true);
+        }
+      }
+    }
   }
 }

```

Reply:
I think this approach is okay. Be sure it is run on multinode. And be sure no leaks occur.